### ISSUE TITLE
Add support for function objects in attachInterrupt

### ISSUE THREAD

**Issue by igrr** [esp8266/Arduino#1925]
Allow something like 

``` c++
attachInterrupt(pin, [](){
    // function body
  }, CHANGE);
```

and

``` c++
class MyClass() {
public:
  MyClass(int pin1, int pin2) {
    attachInterrupt(pin1, std::bind(&MyClass::onPinChange, this, pin1), CHANGE);
    attachInterrupt(pin2, std::bind(&MyClass::onPinChange, this, pin2), CHANGE);
  }

protected:
  void onPinChange(int pin) {
    // handler here
  }
};
```

without breaking usage of attachInterrupt from C code.
##

**Comment by hreintke** [comment]
Like to use this functionality within my application and willing to give it a try creating a PR for this.
Do you already have a basic idea on how you want to have it implemented.
If so, I can base my work on that.

**Comment by neu-rah** [comment]
i've working on something like this, pin change interrupts with an optional cargo, https://github.com/neu-rah/PCINT

**Comment by hreintke** [comment]
See https://github.com/esp8266/Arduino/pull/2745
(#2745)

**Comment by zenmanenergy** [comment]
Wow. Did this ever get into the core? This is a very cool idea.

**Comment by ferchinas** [comment]
Hi:
I am trying to attach a method of my class to attend a Gpio interrupt. But I get the following error:
GwiModules.cpp: 283:122: error: cannot convert 'std::_Bind_helper<false, void (GwiManual::*)(char), GwiManual* const, int>::type {aka std::_Bind<std::_Mem_fn<void (GwiManual::*)(char)>(GwiManual*, int)>}' to 'void (*)()' for argument '2' to 'void attachInterrupt(uint8_t, void (*)(), int)

NodeMCU V1.0
Arduino IDE V1.8.3
Arduino/Esp8266 V2.3


GwiModules.h
```
  class GwiManual : public Controls
  {
  protected:
	  void event(void);							
	  
  public:
	  GwiManual();
	  void enable(void);	 
  };
```
GwiModules.cpp
```
GwiManual::GwiManual() {
	...
}

void GwiManual::event(void) {
	...
}

void GwiManual::enable(void) {
	...
	attachInterrupt(13, std::bind(&GwiManual::event,this), CHANGE
}

```

**Comment by devyte** [comment]
Did you #include "FunctionalInterrupts.h"?

**Comment by ferchinas** [comment]
I did not include it, but thanks to your message I discovered that this information is for the development version.
I can not find how to install it in the IDE, only found for stable version.

**Comment by devyte** [comment]
Read using git version on readthedocs. You have to delete the stable core
dir to make sure the git is picked up.

**Comment by ferchinas** [comment]
Finally get it installed and tested, it works very well, thank you very much for the help.

**Comment by devyte** [comment]
@igrr the relevant PR is already merged. Should this be closed, or left open with label staged-for-release, or something else?

**Comment by tonton81** [comment]
shouldnt you also show the state change not just the pin itself? some people might want to check rising and falling edges, not just changes, if a state was returned the user wouldnt need to call digitalRead when the interrupt fires to know whether the pin changed high or low as well

**Comment by devyte** [comment]
This is already implemented via #2745 .
